File: eventual-consistency-support-core/src/main/java/org/xyattic/eventual/consistency/support/core/aop/support/AnnotationClassOrMethodMatcher.java
package org.xyattic.eventual.consistency.support.core.aop.support;

import org.springframework.aop.ClassFilter;
import org.springframework.aop.support.annotation.AnnotationMethodMatcher;
import org.springframework.core.annotation.AnnotatedElementUtils;

import java.lang.annotation.Annotation;
import java.lang.reflect.Method;

/**
 * @author wangxing
 * @create 2020/11/12
 */
public class AnnotationClassOrMethodMatcher extends AnnotationMethodMatcher {

    private final ClassFilter classFilter;

    public AnnotationClassOrMethodMatcher(Class<? extends Annotation> annotationType) {
        super(annotationType, true);
        this.classFilter = clazz -> AnnotatedElementUtils.isAnnotated(clazz, annotationType);
    }

    @Override
    public boolean matches(Method method, Class<?> targetClass) {
        return super.matches(method, targetClass) || classFilter.matches(targetClass);
    }

}

File: eventual-consistency-support-core/src/main/java/org/xyattic/eventual/consistency/support/core/aop/support/AnnotationClassOrMethodMatcher.kt
package org.xyattic.eventual.consistency.support.core.aop.support

import org.springframework.aop.ClassFilter
import org.springframework.aop.support.annotation.AnnotationMethodMatcher
import org.springframework.core.annotation.AnnotatedElementUtils
import java.lang.reflect.Method

/**
 * @author wangxing
 * @create 2020/11/12
 */
open class AnnotationClassOrMethodMatcher(annotationType: Class<out Annotation>)
    : AnnotationMethodMatcher(annotationType, true) {

    private val classFilter: ClassFilter

    override fun matches(method: Method, targetClass: Class<*>): Boolean {
        return super.matches(method, targetClass) || classFilter.matches(targetClass)
    }

    init {
        classFilter = ClassFilter { clazz: Class<*> -> AnnotatedElementUtils.isAnnotated(clazz, annotationType) }
    }
}